### PR TITLE
Updating the R autograder

### PR DESCRIPTION
**Pull request by eddelbuettel**
Changes are almost exclusively white-space or 'labeling' as eg the underlying container can now be referred to as 'rocker/r2u' rather than 'eddelbuettel/r2u'.

Similarly, as r2u 'maps' CRAN packages to the Ubuntu binaries it provides we can refer to them by their CRAN package name which makes this easier for all R users (and not just those used to Debian/Ubuntu naming schemes).

Only one new package: duckdb

### CODE REVIEW

**Comment by nwalters512** [comment]
I'm going to quickly close and re-open this in the hope that that allows the image builds to get pushed to Docker Hub for testing and size comparisons. Unfortunately our build tooling doesn't automatically support that for contributions from forks.

**Comment by eddelbuettel** [comment]
Sure.  In case it helps, my local test build came in at 1.36gb (relative to 829mb on the underlying image -- note how images 2 to 7 have the same image id)

```sh
edd@rob:~/git/prairielearn/graders/r(master)$ docker images | head -8
REPOSITORY                               TAG             IMAGE ID       CREATED          SIZE
local-pl-r-grader                        latest          98108b204eb7   30 minutes ago   1.36GB
rocker/r2u                               24.04           fa8c36a7520e   21 hours ago     829MB
rocker/r2u                               latest          fa8c36a7520e   21 hours ago     829MB
rocker/r2u                               noble           fa8c36a7520e   21 hours ago     829MB
eddelbuettel/r2u                         24.04           fa8c36a7520e   21 hours ago     829MB
eddelbuettel/r2u                         latest          fa8c36a7520e   21 hours ago     829MB
eddelbuettel/r2u                         noble           fa8c36a7520e   21 hours ago     829MB
edd@rob:~/git/prairielearn/graders/r(master)$ 
```

File: graders/r/Dockerfile
Comment (nwalters512) [inline graders/r/Dockerfile:34]: It looks like duckdb is already installed? Or is that the R library, and this comment is referencing a system package?

```suggestion
# SQL lectures. (Adding a duckdb binary may be an option too.)
```

Comment (eddelbuettel) [reply graders/r/Dockerfile:34]: Yes there is 'duckdb' the R package from CRAN (installed) and there is 'duckdb' the all-in-one binary (cf 'sqlite') which one could get from [here](https://duckdb.org/docs/installation/?version=stable&environment=cli&platform=linux&download_method=direct&architecture=x86_64) -- but because this a grader is possibly not as compelling unless one does a lot of SQL.

File: graders/r/Dockerfile
Comment (eddelbuettel) [reply graders/r/Dockerfile:1]: That'ss how would write it, wrote it.  The indentation is _consistent within each block_ of apt packages, and then CRAN packages.

image

Your repo. Do what you need to do if that bugs you.

Comment (eddelbuettel) [reply graders/r/Dockerfile:1]: As for the earlier question about size, I see plain rocker/r2u going from 741mb for 22.04 to 829mb for 24.04 so there is some, then there likely is 'some' increase within each package we have -- and there is about 41mb on my machine for R package 'duckdb'.

So yes, sounds unavoidable -- usual balance of 'enough useful stuff keeping it above a minimal "slim" setup.

Comment (eddelbuettel) [reply graders/r/Dockerfile:1]: > LGTM, will approve after discussing the formatting.

Just make whatever edits you need to make.

Comment (eddelbuettel) [reply graders/r/Dockerfile:1]: As for indentation a little different at the top, yes because
- we changed from `apt-get` to `apt`
- we changed from `r-cran-foo` to `foo` for all the R packages

so thise lines are shorter. I left the ones near the bottom to not add more whitespace changes.  I have not strong feelings about indenting 'relatve to `RUN .` ie four space or via a full tab.  I seem to use both.

**Comment by eddelbuettel** [comment]
@nwalters512 To resolve this I need either a) actionable feedback from you or b) simply let you alter to fit your requirements.  

No need to overcomplicate this. I have no stakes in indentations wars.

File: graders/r/Dockerfile
Comment (nwalters512) [reply graders/r/Dockerfile:1]: For clarity, this is what I mean by "inconsistent indentation" (this is a screenshot from GitHub's "files changed" tab):

Screenshot 2024-11-07 at 09 12 03

It does indeed look like this is caused by mixing tabs and spaces, which I can see if I view this file in my editor. I'll push a change to consistently use spaces like we do elsewhere.

Comment (eddelbuettel) [reply graders/r/Dockerfile:1]: Ohhhhh I see now!  That is indeed horrible -- and of course not what I see in Emacs, but entirely possibly as this was a file written by multiple authors.

I can run `C-u M-x tabify` to get the tabs out.

File: graders/r/Dockerfile
Comment (eddelbuettel) [reply graders/r/Dockerfile:34]: Correct. `library(duckdb)` in R runs against an R package that has the object code baked in and does not in any way or shape or form rely on command-line tool.

File: graders/r/Dockerfile
Comment (nwalters512) [reply graders/r/Dockerfile:1]: For context on why I brought this up: my brain was trying to pattern-match this against similar formatting I've seen in some Python requirements files, where such indentation indicates that, for instance, `duckdb` is a requirement for `doparallel`. It's now clear that this isn't the case and that this is just a difference in the width that your editor vs. GitHub's UI renders tabs.

Comment (eddelbuettel) [reply graders/r/Dockerfile:1]: Done in 21b4eeff249312f9b93b6be2d2206f3a4e3a24c2

File: graders/r/Dockerfile
Comment (nwalters512) [inline graders/r/Dockerfile:92]: Tiny nit since this got caught up in the removal of tabs: indentation isn't actually required here, so let's just drop it.

We're maybe going down a rabbithole here, but: is there a reason this image goes out of its way to set up a git user? When will the grader ever make commits to a repository?

```suggestion
        && echo "name = Autograding User" >> /home/ag/.gitconfig \
        && echo "email = ag@nowhere" >> /home/ag/.gitconfig \
```

**Review by nwalters512 — COMMENTED** [review]
Thanks for resolving the indentation issue!

File: graders/r/Dockerfile
Comment (eddelbuettel) [reply graders/r/Dockerfile]: Because 'Data Science Programming Methods' covers shell, awk, sed, ... we also hit git which we test the same way via 'quasi unit tests' which the grader runs and accumulates.   So we need the ability to run git from the shell and to compare its output and I think I concluded a while back that for that we do need a git config.

Comment (nwalters512) [reply graders/r/Dockerfile]: Huh, interesting. Thanks for the explanation!

Comment (eddelbuettel) [reply graders/r/Dockerfile]: Sorry hit 'resolved' here by accident. That is your call, of course.

**Comment by nwalters512** [comment]
CI is failing because of a change in a transitive dependency. This should be resolved once https://github.com/PrairieLearn/PrairieLearn/pull/10908 lands.

**Comment by eddelbuettel** [comment]
Let me know if you want me to hit 'rebase' here.  It should not matter because I presume you will squash-merge.

File: graders/r/Dockerfile
Comment (jonatanschroeder) [reply graders/r/Dockerfile]: Have you considered using `git config -f /home/ag/.gitconfig user.email ag@nowhere` instead?

Comment (eddelbuettel) [reply graders/r/Dockerfile]: Instead of the redirect?

Comment (jonatanschroeder) [reply graders/r/Dockerfile]: That command will set up the config file in the appropriate format so you don't need to worry about writing directly to the file. This is used elsewhere in PL too:
https://github.com/PrairieLearn/PrairieLearn/blob/613df286061077064dcf44574bd29d4f767c83a6/Dockerfile#L50-L52

This is a nit, of course, feel free to ignore it.

Comment (eddelbuettel) [reply graders/r/Dockerfile]: I like that approach of writing the config file via calls to `git config`.  In fact, my slides show the same.   Then again, the exiting code should be robust...  

And I could also copy an existing file in.

**Review by nwalters512 — APPROVED** [review]

**Comment by eddelbuettel** [comment]
Yay!  Thanks for merging.

**Comment by ZacWarham** [comment]
Thanks for getting this sorted everyone 🎉

**Pull request merged**
